incubator/group/keeper.go

```diff
@@ -0,0 +1,247 @@
+	k.groupMemberByGroupIndex = orm.NewUInt64Index(groupMemberTableBuilder, GroupMemberByGroupIndexPrefix, func(val interface{}) ([]uint64, error) {
+		group := val.(*GroupMember).Group
+		return []uint64{uint64(group)}, nil
+	})
+	k.groupMemberByMemberIndex = orm.NewIndex(groupMemberTableBuilder, GroupMemberByMemberIndexPrefix, func(val interface{}) ([]orm.RowID, error) {
+		member := val.(*GroupMember).Member
+		return []orm.RowID{member.Bytes()}, nil
+	})
+	k.groupMemberTable = groupMemberTableBuilder.Build()
+
+	//
+	// Group Account Table
+	//
+	groupAccountTableBuilder := orm.NewNaturalKeyTableBuilder(GroupAccountTablePrefix, storeKey, &GroupAccountMetadataBase{}, orm.Max255DynamicLengthIndexKeyCodec{})
+	k.groupAccountByGroupIndex = orm.NewUInt64Index(groupAccountTableBuilder, GroupAccountByGroupIndexPrefix, func(value interface{}) ([]uint64, error) {
+		group := value.(*GroupAccountMetadataBase).Group
+		return []uint64{uint64(group)}, nil
+	})
+	k.groupAccountByAdminIndex = orm.NewIndex(groupAccountTableBuilder, GroupAccountByAdminIndexPrefix, func(value interface{}) ([]orm.RowID, error) {
+		admin := value.(*GroupAccountMetadataBase).Admin
+		return []orm.RowID{admin.Bytes()}, nil
+	})
+	k.groupAccountTable = groupAccountTableBuilder.Build()
+
+	//
+	// ProposalBase Table
+	//
+	ProposalBaseTableBuilder := orm.NewAutoUInt64TableBuilder(ProposalBaseTablePrefix, ProposalBaseTableSeqPrefix, storeKey, &ProposalBase{})
+	k.ProposalBaseByGroupAccountIndex = orm.NewIndex(ProposalBaseTableBuilder, ProposalBaseByGroupAccountIndexPrefix, func(value interface{}) ([]orm.RowID, error) {
+		account := value.(*ProposalBase).GroupAccount
+		return []orm.RowID{account.Bytes()}, nil
+
+	})
+	k.ProposalBaseByProposerIndex = orm.NewIndex(ProposalBaseTableBuilder, ProposalBaseByProposerIndexPrefix, func(value interface{}) ([]orm.RowID, error) {
+		proposers := value.(*ProposalBase).Proposers
+		r := make([]orm.RowID, len(proposers))
+		for i := range proposers {
+			r[i] = proposers[i].Bytes()
+		}
+		return r, nil
+	})
+	k.ProposalBaseTable = ProposalBaseTableBuilder.Build()
+
+	//
+	// Vote Table
+	//
+	voteTableBuilder := orm.NewNaturalKeyTableBuilder(VoteTablePrefix, storeKey, &Vote{}, orm.Max255DynamicLengthIndexKeyCodec{})
+	k.voteByProposalBaseIndex = orm.NewUInt64Index(voteTableBuilder, VoteByProposalBaseIndexPrefix, func(value interface{}) ([]uint64, error) {
+		return []uint64{uint64(value.(*Vote).Proposal)}, nil
+	})
+	k.voteByVoterIndex = orm.NewIndex(voteTableBuilder, VoteByVoterIndexPrefix, func(value interface{}) ([]orm.RowID, error) {
+		return []orm.RowID{value.(*Vote).Voter.Bytes()}, nil
+	})
+	k.voteTable = voteTableBuilder.Build()
+
+	return k
+}
+
+func (k Keeper) CreateGroup(ctx orm.HasKVStore, admin sdk.AccAddress, members []Member, comment string) (GroupID, error) {
+	id := k.groupSeq.NextVal(ctx)
+	var groupID = GroupID(id)
+	err := k.groupTable.Create(ctx, orm.EncodeSequence(id), &GroupMetadata{
```

Comment:
I have replaced the `AutoUInt64Table` with the `orm.Table` for Gas optimization. The groupID stored with the entity costs us 1000 Gas extra otherwise for reading the sequence twice. Please see the commit before. Alternatives would be to use a sequence and natural table but that won't give us the proper API or introduce a `BeforeSave` Interceptor that can be used to modify the entity and set the groupID. WDYT?

Reply:
keep -> new issue to address this